src/goto-instrument/contracts/dynamic-frames/dfcc_cfg_info.cpp

```diff
@@ -0,0 +1,899 @@
+          tracked.insert(ident);
+      }
+    }
+  }
+  return tracked;
+}
+
+struct contract_clausest
+{
+  exprt invariant_expr;
+  assignst assigns;
+  exprt::operandst decreases_clauses;
+};
+
+/// Generate defaults for all contract clauses of the loop with ID `loop_id` if
+/// at least one type of clause is defined
+static struct contract_clausest default_loop_contract_clauses(
+  const dfcc_loop_nesting_grapht &loop_nesting_graph,
+  const std::size_t loop_id,
+  const irep_idt &function_id,
+  local_may_aliast &local_may_alias,
+  message_handlert &message_handler,
+  const namespacet &ns)
+{
+  messaget log(message_handler);
+
+  const auto loop = loop_nesting_graph[loop_id];
+  // Process loop contract clauses
+  exprt::operandst invariant_clauses = get_invariants(loop.latch);
+  exprt::operandst assigns_clauses = get_assigns(loop.latch);
+
+  // initialise defaults
+  struct contract_clausest result;
+  result.invariant_expr = nil_exprt();
+  result.decreases_clauses = get_decreases(loop.latch);
+
+  // Generate defaults for all clauses if at least one type of clause is defined
+  if(
+    !invariant_clauses.empty() || !result.decreases_clauses.empty() ||
+    !assigns_clauses.empty())
+  {
+    if(invariant_clauses.empty())
+    {
+      // use a default invariant if none given.
+      result.invariant_expr = true_exprt{};
+      // assigns clause is missing; we will try to automatic inference
+      log.warning() << "No invariant provided for loop " << function_id << "."
+                    << loop.latch->loop_number << " at "
+                    << loop.head->source_location()
+                    << ". Using 'true' as a sound default invariant. Please "
+                       "provide an invariant the default is too weak."
+                    << messaget::eom;
+    }
+    else
+    {
+      // conjoin all invariant clauses
+      result.invariant_expr = conjunction(invariant_clauses);
+    }
+
+    // unpack assigns clause targets
+    if(!assigns_clauses.empty())
+    {
+      for(auto &operand : assigns_clauses)
+      {
+        result.assigns.insert(operand);
+      }
+    }
+    else
+    {
+      // infer assigns clause targets if none given
+      auto inferred = dfcc_infer_loop_assigns(
+        local_may_alias, loop.instructions, loop.head->source_location(), ns);
+      log.warning() << "No assigns clause provided for loop " << function_id
+                    << "." << loop.latch->loop_number << " at "
+                    << loop.head->source_location() << ". The inferred set {";
+      bool first = true;
+      for(const auto &expr : inferred)
+      {
+        if(!first)
+        {
+          log.status() << ", ";
+        }
+        first = false;
+        log.status() << format(expr);
+      }
+      log.status() << "} might be incomplete or imprecise, please provide an "
+                      "assigns clause if the analysis fails."
+                   << messaget::eom;
```

Comment:
You might want to use `log.conditional_output(log.status(), [&](messaget::mstreamt &mstream) { ... < the above code > }` to make clear that all of this is just about logging. (I usually recommend this for debug output to improve performance; here it's more about readability.)

Reply:
This should actually be all printed as a warning.